File: 18. Backend Scaling and Performance Engineering-part-1/code/typescript/example_1.ts
export function percentile(latencies: number[], p: number): number {
    if (latencies.length === 0) return 0;
    const sorted = [...latencies].sort((a, b) => a - b);
    const rank = (p / 100.0) * (sorted.length - 1);
    const idx = Math.ceil(rank);
    return sorted[idx];
}

const samples = [12, 15, 22, 48, 95, 110, 340, 890, 1200, 4800];
console.log("P50: ", percentile(samples, 50));
console.log("P90: ", percentile(samples, 90));
console.log("P99: ", percentile(samples, 99));

File: 18. Backend Scaling and Performance Engineering-part-1/code/rust/example_1.rs
pub fn percentile(latencies: &[std::time::Duration], p: f64) -> std::time::Duration {
    if latencies.is_empty() {
        return std::time::Duration::from_millis(0);
    }
    let mut sorted = latencies.to_vec();
    sorted.sort();
    let rank = (p / 100.0) * (sorted.len() - 1) as f64;
    let idx = rank.ceil() as usize;
    sorted[idx]
}

fn main() {
    let samples = vec![
        std::time::Duration::from_millis(12), std::time::Duration::from_millis(15),
        std::time::Duration::from_millis(22), std::time::Duration::from_millis(48),
        std::time::Duration::from_millis(95), std::time::Duration::from_millis(110),
        std::time::Duration::from_millis(340), std::time::Duration::from_millis(890),
        std::time::Duration::from_millis(1200), std::time::Duration::from_millis(4800),
    ];
    println!("P50: {:?}", percentile(&samples, 50.0));
    println!("P90: {:?}", percentile(&samples, 90.0));
    println!("P99: {:?}", percentile(&samples, 99.0));
}
